crates/ra_hir/src/ty/infer.rs

```diff
@@ -286,7 +285,9 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
             | (Ty::Infer(InferTy::IntVar(tv)), other)
             | (other, Ty::Infer(InferTy::IntVar(tv)))
             | (Ty::Infer(InferTy::FloatVar(tv)), other)
-            | (other, Ty::Infer(InferTy::FloatVar(tv))) => {
+            | (other, Ty::Infer(InferTy::FloatVar(tv)))
+                if !is_never(other) =>
```

Comment:
@flodiebold, @SomeoneToIgnore am I correct that we still want to remove this special case from unifcation before merging?

Reply:
Yes, that's the idea.
I've made a small attempt to do it the way proposed, but it did not work instantly alas.
I'll start investigating things in-depth next week, sorry for letting it hang for so long.

Reply:
I've investigated the issues and wrote the comment to the discussion above.

What bothers me the most is the first item from that comment's list of issues: other items seem to be easy to fix, but that one requires quite many changes besides the ones that are already done.
If I'm not wrong about it, would it make sense to merge the current code now so that somebody more experienced with the inference logic can refactor it later while the users can have their types inferred correctly still?
We can add an optional FIXME and a follow-up issue, if it helps :)

Reply:
My worry is that this will actually break the inference in other cases (e.g. when something actually needs to unify with `!`; although I'd admit that in total, it would probably increase the number of correctly inferred cases). I really appreciate you trying to apply my suggestions though. I'll see whether I can find a good solution for the issues you listed (probably this evening).